### DISCUSSION TITLE
How to convert output to Gin data without struct definition ?

### DISCUSSION THREAD

**Discussion by ArenaSu** [cue-lang/cue#1932] [Q&A] [answered]
I create a test.cue below: 

```
output: {
	total: parameter.total
	applications: [
      for a in parameter.applications {
          name: a.metadata.name
          displayName: a.spec.displayName
          type: a.spec.type
          creationTimestamp: a.metadata.creationTimestamp
          message: string | *""
          operation: string | *""
          operationStatus: string | *""
          firstExecuteTime: string | *""
          lastExecuteTime: string | *""
          if a.status != _|_ & a.status.operationStatus != _|_ {
          	  phase: a.status
              operation:  a.status.operationStatus.type
              operationStatus: a.status.operationStatus.status
              firstExecuteTime: a.status.operationStatus.firstExecuteTime
              lastExecuteTime: a.status.operationStatus.lastExecuteTime
           }
      }
  ]
}
Parameter: {
...
}
```

And i got ouput by code below

```
o := val.LookupPath(cue.ParsePath("output"))
data, err := v.MarshalJSON()
```
I pass data to Gin ctx.JSON, but I got "eyJ0b3RhbCI6MCwiYXBwbGljYXRpb25zIjpbXX0=" from response body. 

How can i fix it ?

**Comment by ArenaSu** [comment]
@mvdan Could you help me as i'm stuck on this question ？

**Comment by mvdan** [comment, marked as answer]
This all depends on what your whole program looks like. What you showed isn't enough to see what's going on. Perhaps try sharing a full reproducer via https://github.com/cue-lang/cue/wiki/Creating-test-or-performance-reproducers, which would do something like `go run` with a Go program and some CUE files.

But also, that response looks like base64-encoded bytes. Your value is just empty:

```
$ base64 -d <<<'eyJ0b3RhbCI6MCwiYXBwbGljYXRpb25zIjpbXX0'
{"total":0,"applications":[]}
```

**Reply by ArenaSu** [reply]
> This all depends on what your whole program looks like. What you showed isn't enough to see what's going on. Perhaps try sharing a full reproducer via https://github.com/cue-lang/cue/wiki/Creating-test-or-performance-reproducers, which would do something like `go run` with a Go program and some CUE files.
> 
> But also, that response looks like base64-encoded bytes. Your value is just empty:
> 
> ```
> $ base64 -d <<<'eyJ0b3RhbCI6MCwiYXBwbGljYXRpb25zIjpbXX0'
> {"total":0,"applications":[]}
> ```

My raw data is just below: 
```
{"total":0,"applications":[]}
```

It's right and the key point is Gin render []byte data to base64 string.Maybe i should create an issue in Gin projcet.It's not cuelang question.

**Reply by roman-mazur** [reply]
It's not a CUE question indeed. Most likely, you can call [`ctx.Data`](https://pkg.go.dev/github.com/gin-gonic/gin#Context.Data) instead of `JSON` to get the expected result.
```
ctx.Data(200, "application/json", data)
```

**Reply by ArenaSu** [reply]
Thanks, i use ctx.String and set json header.
